sale_order_discount_invoicing/models/account_move.py

```diff
@@ -0,0 +1,154 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import copy
+
+from odoo import _, api, models
+from odoo.exceptions import ValidationError
+from odoo.tools import float_compare
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    @api.constrains("line_ids")
+    def _check_split_lines_constrains(self):
+        """
+        We want to be sure split lines are still both together
+        and with same quantities.
+
+        Considering couple of invoice lines with sale line links.
+        """
+        for move in self:
+            split_lines = move.line_ids.filtered("is_split_line")
+            for split_line in split_lines:
+                sale_line_ids = split_line.sale_line_ids
+                quantity = split_line.quantity
+                other_split_lines = split_lines - split_line
+                couple_split_line = other_split_lines.filtered(
+                    lambda l: l.sale_line_ids == sale_line_ids
+                )
+                if not (
+                    len(couple_split_line) == 1
+                    and couple_split_line.quantity == quantity
+                ):
+                    raise ValidationError(
+                        _(
+                            "The line you have deleted/modified is linked to "
+                            "another one in the Sale Order.\n"
+                            "Please delete/modify both lines."
+                        )
+                    )
+
+    def _can_process_discount_line_vals(self, invoice_line_vals):
+        # To allow to override it in specific
+        return True
+
+    def _get_discount_line_values(
+        self, invoice_line_vals, price_unit, origin_price_unit
+    ):
+        product_id = invoice_line_vals["product_id"]
+        discount_line = invoice_line_vals.copy()
+        product = self.env["product.product"].browse(product_id)
+        discount_account = product.product_tmpl_id.get_product_accounts(
+            fiscal_pos=self.fiscal_position_id
+        )["discount"]
+        discount_line.update(
+            {
+                "account_id": discount_account.id,
+                "price_unit": price_unit - origin_price_unit,
+                "name": _("Discount on %s") % discount_line["name"],
+                "sale_line_ids": invoice_line_vals.get("sale_line_ids"),
+                "is_split_line": True,
+                "is_split_discount_line": True,
+            }
+        )
+        return discount_line
+
+    def _process_discount_line_vals(self, invoice_line_vals):
+        """if the invoice line has a slashed price, split it in two, one for the full
+        price and one for the discount. Otherwise keep it"""
+        if "origin_price_unit" in invoice_line_vals:
+            origin_price_unit = invoice_line_vals.pop("origin_price_unit")
+            if not self._can_process_discount_line_vals(invoice_line_vals):
+                yield invoice_line_vals
+            else:
```

Comment:
From what I understand, the test `_can_process_discount_line_vals` could be at the beginning of the function : 
```suggestion
            if not self._can_process_discount_line_vals(invoice_line_vals):
                yield invoice_line_vals
            else:
                if "origin_price_unit" in invoice_line_vals:
                    origin_price_unit = invoice_line_vals.pop("origin_price_unit")
```

And if you do that, you may not need to add the function `_can_process_discount_line_vals` here for the specific later, you could directly remove the test here : 
```python
         if "origin_price_unit" in invoice_line_vals:
                    origin_price_unit = invoice_line_vals.pop("origin_price_unit")
                    price_unit = invoice_line_vals["price_unit"]
                    ...
```

And then in your specific, make : 
```python
def _process_discount_line_vals(self, invoice_line_vals):
        if not self._can_process_discount_line_vals(invoice_line_vals):
                yield invoice_line_vals
        else:
                return super()._process_discount_line_vals
```